ext/tbb/tbb_spin_rw_mutex.h

```diff
@@ -0,0 +1,291 @@
+ *      It will back off and eventually yield the CPU when writer is holding a
+ *      lock until the lock becomes available again.
+ *
+ *  Readers/Writers ratio (-r) and Pure readers (-m):
+ *
+ *      - 'rw_mask' variable defines the ratio between readers and writers per
+ *      thread. It is controlled using log2_ratio variable, cmdline args -r.
+ *
+ *      - Given the ratio, a thread will perform that many 'read_acquire' and
+ *      'read_release' calls and then it will do one 'write_acquire' and
+ *      'write_release'. And then if more work to be done, repeat.
+ *
+ *      For a thread:
+ *
+ *      num readers
+ *      ----------- = 2^(log2_ratio) - 1;
+ *      num writers
+ *
+ *       > log2_ratio of  0 means all writers
+ *       > log2_ratio of ~0 means all readers
+ *       > default log2_ratio is 6 e.g 63 reads per write.
+ *
+ *      - Pure readers are CPUs which will never perform a write acq/rel. The
+ *        cmdline arg is a bit mask e.g. 0x8 will make 4th cpu (cpu id: 0x3)
+ *        a pure reader. Default is 0x0 e.g. no pure readers.
+ *
+ */
+
+#ifndef __TBB_spin_mutex_H
+#define __TBB_spin_mutex_H
+
+#define initialize_lock(lock, threads) tbb_init_locks(lock, threads)
+#define parse_test_args(args, argc, argv) tbb_parse_args(args, argc, argv)
+
+#include "tbb.h"
+
+#define WRITER          1
+#define WRITER_PENDING  2
+#define READERS         ~(WRITER | WRITER_PENDING)
+#define ONE_READER      4
+#define BUSY            (WRITER | READERS)
+
+unsigned long log2_ratio = 0;
+unsigned long rw_mask = 0;
+unsigned long reader_cpu_mask = 0;
+
+typedef struct {
+    unsigned long c;
+    uint8_t pure_reader;
+} __attribute__((aligned(64))) rw_count_t;
+
+rw_count_t *rw_counts;
+
+inline uint8_t is_writer(unsigned long i, uint8_t val) {
+    if (rw_counts[i].pure_reader)
+        return 0;
+    rw_counts[i].c += val;
+    return !(rw_counts[i].c & rw_mask);
+}
+
+void tbb_print_usage() {
+    fprintf(stderr, " tbb_spin_rw_mutex\n");
+    fprintf(stderr, "\t[-h print this msg]\n");
+    fprintf(stderr, "\t[-r reader/writer log ratio, default: 6 (2^(6)-1 readers per writer)]\n");
+    fprintf(stderr, "\t[-m pure reader cpu mask, default: 0x0 (no pure readers)]\n");
+}
+
+void tbb_parse_args(test_args unused, int argc, char** argv) {
+    int i = 0;
+
+    log2_ratio = 6;
+    reader_cpu_mask = 0x0;
+
+    while ((i = getopt(argc, argv, "hr:m:")) != -1)
+    {
+        switch (i) {
+          case 'r':
+            log2_ratio = strtoul(optarg, (char **) NULL, 10);
+            if (log2_ratio >= 64) {
+                fprintf(stderr, "tbb_spin_rw_mutex: -r can not be >= 64\n");
+                exit(1);
+            }
+            break;
+          case 'm':
+            if (!strncmp(optarg, "0x", 2))
+                reader_cpu_mask = strtoul(optarg, (char **) NULL, 16);
+            else
+                reader_cpu_mask = strtoul(optarg, (char **) NULL, 10);
+            if (errno == ERANGE) {
```

Comment:
Should we check strtoul() return value first then global errno? Because errno may accidentally have previous value of ERANGE.

Reply:
Hi Zoy,

You're right.  Unfortunately the return value of strtoul varies greatly depending on what type of error is encountered and could include 0 and 0xFFFF...FFFF both of which are potential valid and useful mask values.  Perhaps the best thing to do here is to simply take the output of strtoul at face value and not check errno?